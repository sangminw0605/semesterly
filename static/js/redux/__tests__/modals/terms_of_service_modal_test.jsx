/*
Copyright (C) 2017 Semester.ly Technologies, LLC

Semester.ly is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Semester.ly is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
*/
import React from 'react';
import { render } from '@testing-library/react';
import thunk from 'redux-thunk';
import Provider from 'react-redux/src/components/Provider';
import configureMockStore from 'redux-mock-store';
import {
  tosModalFixture,
  userInfoFixture,
} from '../../__fixtures__/terms_of_service_modal.fixture';
import { handleAgreement } from '../../actions/user_actions';
import * as ActionTypes from '../../constants/actionTypes';
import TermsOfServiceModalContainer from '../../ui/containers/terms_of_service_modal_container';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('TOS Modal', () => {
  it('shows when isVisible is true', () => {
    const store = mockStore({
      termsOfServiceModal: tosModalFixture,
      userInfo: userInfoFixture,
    });

    const { container } = render(
      <Provider store={store}>
        <TermsOfServiceModalContainer />
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });

  it('shows welcome message for new users', () => {
    const newUser = userInfoFixture;
    newUser.data.timeAcceptedTos = null;
    const store = mockStore({
      termsOfServiceModal: tosModalFixture,
      userInfo: userInfoFixture,
    });
    const { container } = render(
      <Provider store={store}>
        <TermsOfServiceModalContainer />
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });

  it('is hidden when isVisible is false', () => {
    const store = mockStore({
      termsOfServiceModal: { ...tosModalFixture, isVisible: false },
      userInfo: userInfoFixture,
    });
    const { container } = render(
      <Provider store={store}>
        <TermsOfServiceModalContainer />
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});

describe('HandleAgreement correctly triggers tosMODAL', () => {
  it('if unaccepted', () => {
    const store = mockStore({
      termsOfServiceModal: tosModalFixture,
      userInfo: userInfoFixture,
    });

    const currentUser = {
      timeAcceptedTos: null,
      isLoggedIn: true,
    };

    store.dispatch(handleAgreement(currentUser, Date.now()));
    const expectedActions = store.getActions();
    expect(expectedActions[0]).toEqual({ type: ActionTypes.TRIGGER_TOS_MODAL });
  });

  it('if accepted but outdated', () => {
    const store = mockStore({
      termsOfServiceModal: tosModalFixture,
      userInfo: userInfoFixture,
    });

    const currentUser = {
      timeAcceptedTos: 0,
      isLoggedIn: true,
    };

    store.dispatch(handleAgreement(currentUser, Date.parse(1)));
    const expectedActions = store.getActions();
    expect(expectedActions[0]).toEqual({ type: ActionTypes.TRIGGER_TOS_MODAL });
  });

  it('empty if logged in and accepted', () => {
    const store = mockStore({
      termsOfServiceModal: tosModalFixture,
      userInfo: userInfoFixture,
    });

    const currentUser = {
      timeAcceptedTos: 2,
      isLoggedIn: true,
    };

    store.dispatch(handleAgreement(currentUser, Date.parse(1)));
    const expectedActions = store.getActions();
    expect(expectedActions).toEqual([]);
  });
});
