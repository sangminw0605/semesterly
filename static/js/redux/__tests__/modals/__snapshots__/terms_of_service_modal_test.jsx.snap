// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TOS Modal is hidden when isVisible is false 1`] = `<div />`;

exports[`TOS Modal shows welcome message for new users 1`] = `
<div>
  <span>
    <div
      style="z-index: 1050; position: fixed; width: 100%; top: 50%; left: 50%;"
      class="terms-of-service-modal max-modal"
    >
      <div
        tabindex="-1"
        style="margin: 0px; background-color: white;"
      >
        <div
          class="tos-modal-container"
        >
          <h1>
            Terms of Service and Privacy Policy
          </h1>
          <h3>
            Welcome to Semester.ly! Please take a look at our Terms of Service and Privacy Policy before getting started:
          </h3>
          <div>
            <a
              href="/termsofservice"
              target="_blank"
              rel="noopener noreferrer"
              class="legal-links"
            >
              Terms of Service
              <i
                class="fa fa-external-link"
              />
            </a>
            <a
              href="/privacypolicy"
              target="_blank"
              rel="noopener noreferrer"
              class="legal-links"
            >
              Privacy Policy
              <i
                class="fa fa-external-link"
              />
            </a>
          </div>
          <button
            class="accept-tos-btn"
          >
            <i
              class="fa fa-check"
            />
            <span>
              I accept the Terms of Service
            </span>
          </button>
          <p
            class="method-details"
          >
            You must accept the new Terms of Service to continue using Semester.ly.
          </p>
        </div>
      </div>
    </div>
    <div
      style="position: fixed; top: 0px; right: 0px; bottom: 0px; left: 0px; z-index: 1040; background-color: rgb(55, 58, 71);"
    />
  </span>
</div>
`;

exports[`TOS Modal shows when isVisible is true 1`] = `
<div>
  <span>
    <div
      style="z-index: 1050; position: fixed; width: 100%; top: 50%; left: 50%;"
      class="terms-of-service-modal max-modal"
    >
      <div
        tabindex="-1"
        style="margin: 0px; background-color: white;"
      >
        <div
          class="tos-modal-container"
        >
          <h1>
            Terms of Service and Privacy Policy
          </h1>
          <h3>
            Our Terms of Service and Privacy Policy have been updated. Please review them here:
          </h3>
          <div>
            <a
              href="/termsofservice"
              target="_blank"
              rel="noopener noreferrer"
              class="legal-links"
            >
              Terms of Service
              <i
                class="fa fa-external-link"
              />
            </a>
            <a
              href="/privacypolicy"
              target="_blank"
              rel="noopener noreferrer"
              class="legal-links"
            >
              Privacy Policy
              <i
                class="fa fa-external-link"
              />
            </a>
          </div>
          <button
            class="accept-tos-btn"
          >
            <i
              class="fa fa-check"
            />
            <span>
              I accept the Terms of Service
            </span>
          </button>
          <p
            class="method-details"
          >
            You must accept the new Terms of Service to continue using Semester.ly.
          </p>
        </div>
      </div>
    </div>
    <div
      style="position: fixed; top: 0px; right: 0px; bottom: 0px; left: 0px; z-index: 1040; background-color: rgb(55, 58, 71);"
    />
  </span>
</div>
`;
